### DISCUSSION TITLE
Is there any way to access `context` provided by react-hook-form inside zod's `superRefine`?

### DISCUSSION THREAD

**Discussion by DiegoSaturno** [react-hook-form/react-hook-form#11403] [Q&A]
Context:

I have a validation on my form that depends on a calculated value, provided by an xstate selector. I tried to access it through zod's `RefinementContext` but wasn't able to do so.

Is there any way I can get this value inside `superRefine`? If not, is there any way I can write a custom resolver and merge it with `zodResolver`?

The code looks like this:
```tsx
const mySchema = z.object({
   ...
}).superRefine((data, context) => {
  // I want to get `myCalculatedValue` here
});
type MySchema = z.infer<typeof mySchema>;

const calculatedValue = useSelector(actor, selector);

const methods = useForm<MySchema>({
  mode: "all",
  resolver: zodResolver(mySchema),
  context: {
    myCalculatedValue: calculatedValue,
  },
})
```

**Comment by vlanemcev** [comment]
Hi, @DiegoSaturno !

Any progress on this question? I'm trying to figure out how to do this correctly with Zod. Because in Yup library, for example, this approach is really common and easy to use.

**Comment by tunahanbayindir** [comment]
Any updates on this?

**Comment by Unknown47Dream** [comment]
Please add this.

**Comment by youbek** [comment]
Thanks for the library, but is there any updates/workarounds?

**Comment by jeremy-habit** [comment]
up ! pls

**Comment by kpauletti** [comment]
Any update on this? I really need this and i'd prefer not to use yup.

**Comment by landorid** [comment]
This is a must have in my opinion!

**Reply by JPG-2024** [reply]
totally

**Comment by landorid** [comment]
I would love to have this!

**Comment by brianMxBm** [comment]
Surprised this hasn't been added

**Comment by tfazekas** [comment]
Up! Please 🙏

**Comment by fstodulski** [comment]
Guys. I had the same issue.
Went with:
```
  const form = useForm<SubscriptionScheduleSchema>({
    resolver: zodResolver(subscriptionScheduleSchema(amountContext)),
```

where my schema is:
```
export const subscriptionScheduleSchema = (context: {
  minAmount: number;
  maxAmountOther: number;
  maxAmount: number;
}) =>
  newSubscriptionSchema
    .pick({
```

@tfazekas @brianMxBm @landorid 

And it worked!

**Reply by tfazekas** [reply]
@fstodulski 

Yeah... I know, but this case you cant infer type because the schema is dynamic:

```ts
export type SubscriptionScheduleSchemaType = Zod.infer<typeof subscriptionScheduleSchema>;
```

**Reply by fstodulski** [reply]
@tfazekas u can:
`z.infer<ReturnType<typeof termsAndConditionsSchema>>>`

**Reply by fstodulski** [reply]
<img width="529" alt="Screenshot 2025-01-30 at 14 21 56" src="https://github.com/user-attachments/assets/a1d730e5-551a-4e86-9012-ba66ebcbfdba" />
It works like a charm

**Reply by tfazekas** [reply]
@fstodulski 
Aaaah, you're right. Should have tried it out. I’m not really a fan of making the schema a function, as it often encourages teammates to create dynamic schemas.

It works for this case, but consider a scenario where a property is added conditionally based on a context—this would make it unclear to the TypeScript compiler whether that property is present or not.

So in this scenario, if react-hook-form would pass the context to zod and zod.schema's superRefine would receive that context, the schema can remain static.

**Comment by taksuparth** [comment]
@bluebill1049 any plan to add this?

**Reply by vansh-codes** [reply]
@bluebill1049 Exactly when can we expect thisss???

**Reply by idan1idan** [reply]
This is a must! please add a support for this.

**Comment by idan1idan** [comment]
Up!

**Comment by pedrommarquesm** [comment]
crazy how after 1 year this must-have feature isn't still supported

**Comment by arnelyutiga** [comment]
Zod should support context just like Yup

**Comment by annanShao** [comment]
Have any progress?

**Comment by dimitriBouteille** [comment]
Apparently no solution since January 2024 ....

**Comment by BrendanC23** [comment]
https://github.com/colinhacks/zod/issues/4745#issuecomment-3031372210 is a response by the maintainer of Zod which states that "It's very unlikely any context mechanism gets added [to Zod]". The suggested workaround is:

```tsx
function makeSchema(ctx: object) {
  return z.object({
      /// define 
  })
}

type Output = z.infer<ReturnType<typeof makeSchema>>;
```

As long as the schema isn't dynamic and has a consistent input/output types (it only uses the context within a refinement), this is a decent approach. Unless Zod changes to support external context, I don't think there's anything that react-hook-form can do to support this use-case.
